services/openshiftmisc/src/handlers/routeMigration.js

```diff
@@ -0,0 +1,275 @@
+      ? projectOpenShift.openshiftProjectPattern
+          .replace('${branch}', safeStandbyProductionEnvironment)
+          .replace('${project}', safeProjectName)
+      : `${safeProjectName}-${safeStandbyProductionEnvironment}`;
+  } catch (error) {
+    logger.error(`Error while loading information for project ${projectName}`);
+    logger.error(error);
+    throw error;
+  }
+
+  // define the routemigration. the annotation being set to true is what actually triggers the switch
+  const migrateRoutes = (openshiftProject, destinationOpenshiftProject) => {
+    let config = {
+      apiVersion: 'dioscuri.amazee.io/v1',
+      kind: 'RouteMigrate',
+      metadata: {
+        name: openshiftProject,
+        annotations: {
+            'dioscuri.amazee.io/migrate':'true'
+        }
+      },
+      spec: {
+        destinationNamespace: destinationOpenshiftProject,
+        activeEnvironment: safeActiveProductionEnvironment,
+      },
+    };
+
+    return config;
+  };
+
+  // Kubernetes API Object - needed as some API calls are done to the Kubernetes API part of OpenShift and
+  // the OpenShift API does not support them.
+  const dioscuri = new RouteMigration({
+    url: openshiftConsole,
+    insecureSkipTlsVerify: true,
+    auth: {
+      bearer: openshiftToken
+    }
+  });
+  const openshift = new OpenShiftClient.OApi({
+    url: openshiftConsole,
+    insecureSkipTlsVerify: true,
+    auth: {
+      bearer: openshiftToken
+    }
+  });
+
+  // check that the namespaces exist for source and destination before we try and move any routes
+  try {
+    // check source
+    const projectSourceGet = promisify(openshift.projects(openshiftProject).get, { context: openshift.projects(openshiftProject) })
+    projectStatus = await projectSourceGet()
+    logger.info(`${openshiftProject}: Project ${openshiftProject} already exists, continuing`)
+    // check dest
+    const projectDestGet = promisify(openshift.projects(destinationOpenshiftProject).get, { context: openshift.projects(destinationOpenshiftProject) })
+    projectStatus = await projectDestGet()
+    logger.info(`${openshiftProject}: Project ${destinationOpenshiftProject} already exists, continuing`)
+  } catch (err) {
+    // throw error if the namespace doesn't exist
+    logger.error(err)
+    throw new Error
+  }
+
+  // @TODO: this seems a bit silly, might be a better way to do it. but `.patch` on the routemigrates resource fails with,
+  // research says this is because crd is not supported to be patched
+  // `message=the body of the request was in an unknown format - accepted media types include: application/json-patch+json`
+  try {
+    const migrateRoutesDelete = promisify(
+      dioscuri.ns(openshiftProject).routemigrates(openshiftProject).delete
+    );
+    await migrateRoutesDelete({
+      body: {}
+    });
+    await new Promise(resolve => setTimeout(resolve, 10000)); // sleep a bit after deleting
```

Comment:
This makes me nervous... will the timing ever depend on how active or "busy" the k8s cluster/api is? What's the impact of waiting but the the delete isn't "done" yet?

The alternative is to periodically query the API until the crd is actually deleted. An example of that is in the `openshiftremove` service https://github.com/amazeeio/lagoon/blob/master/services/openshiftremove/src/index.js#L228-L236

Reply:
Yeah, this was a bit I was completely unsure on. I'll scope out the referenced code and adjust.